### ISSUE TITLE
Success callbacks for transitions

### ISSUE THREAD

**Issue by brega** [aasm/aasm#236]
Do you think success callbacks 
- defined on transitions
- trigger after the state is persisted

is something worth considering and sending a patch for?

Use case is something like
```ruby
class Order < ActiveRecord::Base
  include AASM
  
  aasm do
    state :unconfirmed 
    state :cash_on_delivery
    state :paid
  
    event :confirm do
      transitions to: :cash_on_delivery, 
                     from: unconfirmed, 
                     success: proc { cash_on_delivery_callbacks },
                     guard: { payment_type == 'Cash on Delivery' }
    
      transitions to: :paid,
                     from: :unconfirmed,
                     success: proc { other_payment_type_callbacks },
                     guard: { payment_type != 'Cash on Delivery' }
    end
  end
end
```

Success callbacks would perform work specific to the transition path and would have to run only after state transition is successful. 

New order of calling would be something like
```
begin
  event           before
  event           guards
  transition      guards
  old_state       before_exit
  old_state       exit
  transition      after
  new_state       before_enter
  new_state       enter
  ...update state...
  transition      success             # if persist successful
  event           success
  old_state       after_exit
  new_state       after_enter
  event           after
rescue
  event           error
end

**Comment by alto** [comment]
Hi @brega 

so, it's basically doing the same thing as event `:success` callbacks, but per transition? What's the benefit from having that, if you could ask for the event name triggered, and the _from_ and _to_ state in the event `:success` callback?

**Comment by alto** [comment]
Hmm, maybe it helps cleaner code... thinking about it.

**Comment by brega** [comment]
Hi @alto 

Yes, thats correct. It does the same as event `:success` callbacks. We have code like

```ruby
event :deliver do
  transitions to: :delivered, from: :packed, success: :packed_state_hooks
  transitions to: :delivered, from: :shipped, success: :shipped_state_hooks
end
```

If we enquire the `from` and `to` from event `:success` callback, we can work, not something I considered . But the callback would be riddled with if else for each transition in there like you noted.

**Comment by brega** [comment]
@alto Anything I can do to further this issue

**Comment by eirvandelden** [comment]
We could certainly use this functionality :+1:

**Comment by chriswoodrich** [comment]
@brega and @eirvandelden I think I'll take a shot at this feature soon. Question for you: should the ```success``` key take a block or the name of a method on the class?

**Comment by eirvandelden** [comment]
@chriswoodrich I'm cool with either. My gut tells me a block could be a more flexible solution, though are success callback is large enough to be it's own method anyhow.
The name of a method gives cleaner code though…

**Comment by chriswoodrich** [comment]
@brega it looks like you've already done most of the work in #239 .  Do you plan to keep going there?  I can pick it up where you left off otherwise.

**Comment by brega** [comment]
@chriswoodrich Not much interest on #239 to be fair and cant remember what else was needed for it. Feel free to take it over.

**Comment by alto** [comment]
Merged #239 now.

**Comment by alto** [comment]
I just released version `4.10.0` including the changes from #239.
